File: python/rizhiyi_mcp/sse_client.py
from __future__ import annotations

import inspect
import json
import re
from dataclasses import dataclass, field
from typing import Any, Callable

import httpx

from .types import HttpClientConfig

_SQL_FENCE_RE = re.compile(r"```spl\s*\n([\s\S]*?)\n```")


@dataclass
class SseEvent:
    event: str
    data: str


@dataclass
class SseStep:
    id: str | None = None
    title: str | None = None
    status: str | None = None
    tool_name: str | None = None


@dataclass
class SseChatResult:
    conversation_id: int | None = None
    conversation: dict[str, Any] | None = None
    spl: str | None = None
    steps: list[SseStep] = field(default_factory=list)


def _parse_sse_block(lines: list[str]) -> SseEvent | None:
    event_name = ""
    data_lines: list[str] = []
    for line in lines:
        if line.startswith("event: "):
            event_name = line[7:].strip()
        elif line.startswith("data: "):
            data_lines.append(line[6:])
    if not event_name or not data_lines:
        return None
    return SseEvent(event=event_name, data="\n".join(data_lines))


def extract_spl_from_markdown(content: str) -> str:
    match = _SQL_FENCE_RE.search(content)
    if match:
        return match.group(1).strip()
    return re.sub(r"```[\s\S]*?```", "", content).strip()


async def request_sse(
    *,
    url: str,
    headers: dict[str, str],
    body: str,
    config: HttpClientConfig,
    timeout_ms: int = 120000,
    on_event: Callable[[SseEvent], Any] | None = None,
) -> SseChatResult:
    """通过 HTTP SSE 流式请求 /api/v3/copilot/chat/ 并聚合结果。"""
    result = SseChatResult()
    request_headers = dict(headers)
    request_headers.setdefault("Accept", "text/event-stream")
    request_headers.setdefault("Cache-Control", "no-cache")
    request_headers.setdefault("Connection", "keep-alive")

    transport = httpx.AsyncHTTPTransport(verify=config.verify_tls)
    async with httpx.AsyncClient(
        base_url=config.base_url,
        headers=config.headers,
        verify=config.verify_tls,
        transport=transport,
        timeout=timeout_ms / 1000.0,
    ) as client:
        async with client.stream("POST", url, headers=request_headers, content=body) as response:
            if response.status_code >= 400:
                error_text = (await response.aread()).decode("utf-8", errors="replace")
                raise httpx.HTTPStatusError(
                    f"SSE request failed ({response.status_code}): {error_text[:500]}",
                    request=response.request,
                    response=response,
                )

            current_lines: list[str] = []
            async for line in response.aiter_lines():
                stripped = line.strip()
                if stripped:
                    current_lines.append(line)
                    continue
                event = _parse_sse_block(current_lines)
                current_lines = []
                if event is None:
                    continue
                _process_event(event, result)
                if on_event is not None:
                    callback_result = on_event(event)
                    if callback_result is not None and inspect.isawaitable(callback_result):
                        await callback_result

    return result


def _process_event(
    event: SseEvent,
    result: SseChatResult,
) -> None:
    if event.event == "DONE":
        try:
            parsed = json.loads(event.data)
            if parsed.get("conversation_id"):
                result.conversation_id = parsed.get("conversation_id")
        except (json.JSONDecodeError, TypeError):
            pass
        return

    if event.event == "META":
        try:
            parsed = json.loads(event.data)
            if parsed.get("conversation"):
                result.conversation = parsed.get("conversation")
        except (json.JSONDecodeError, TypeError):
            pass
        return

    if event.event == "CREATE_STEP":
        try:
            parsed = json.loads(event.data)
            details = parsed.get("details")
            result.steps.append(
                SseStep(
                    id=parsed.get("id"),
                    title=parsed.get("title"),
                    status="CREATED",
                    tool_name=details.get("tool_name") if isinstance(details, dict) else None,
                )
            )
        except (json.JSONDecodeError, TypeError):
            pass
        return

    if event.event == "SET_STATUS":
        try:
            parsed = json.loads(event.data)
            for step in result.steps:
                if step.id == parsed.get("id"):
                    step.status = parsed.get("status")
                    if parsed.get("title"):
                        step.title = parsed.get("title")
                    break
        except (json.JSONDecodeError, TypeError):
            pass
        return

    if event.event == "STEP_OUTPUT":
        try:
            parsed = json.loads(event.data)
            for step in result.steps:
                if step.id == parsed.get("id") and step.tool_name == "send_spl" and parsed.get("content"):
                    result.spl = extract_spl_from_markdown(parsed["content"])
                    break
        except (json.JSONDecodeError, TypeError):
            pass
        return

File: ts/src/sse-client.ts
import http from 'http';
import https from 'https';
import { URL } from 'url';

export interface SseEvent {
    event: string;
    data: string;
}

export interface SseChatResult {
    conversation_id?: number;
    conversation?: Record<string, any>;
    spl?: string;
    steps: Array<{ id: string; title: string; status: string; tool_name?: string }>;
    raw_events: SseEvent[];
}

interface SseClientOptions {
    url: string;
    method?: string;
    headers?: Record<string, string>;
    body?: string;
    timeoutMs?: number;
    onEvent?: (event: SseEvent) => void;
}

export function requestSse(options: SseClientOptions): Promise<SseChatResult> {
    return new Promise((resolve, reject) => {
        const url = new URL(options.url);
        const isHttps = url.protocol === 'https:';
        const transport = isHttps ? https : http;

        const reqHeaders: Record<string, string> = {
            ...options.headers,
            Accept: 'text/event-stream',
            'Cache-Control': 'no-cache',
            Connection: 'keep-alive'
        };

        const reqOptions: https.RequestOptions = {
            hostname: url.hostname,
            port: url.port || (isHttps ? 443 : 80),
            path: url.pathname + url.search,
            method: options.method || 'POST',
            headers: reqHeaders,
            rejectUnauthorized: false
        } as https.RequestOptions;

        const result: SseChatResult = { steps: [], raw_events: [] };
        let buffer = '';
        let finished = false;

        const finish = (err?: Error) => {
            if (finished) return;
            finished = true;
            if (err) {
                reject(err);
            } else {
                resolve(result);
            }
        };

        const req = transport.request(reqOptions, (res) => {
            if (res.statusCode && res.statusCode >= 400) {
                let body = '';
                res.on('data', (chunk: Buffer) => { body += chunk.toString(); });
                res.on('end', () => finish(new Error(`SSE request failed (${res.statusCode}): ${body.slice(0, 500)}`)));
                return;
            }

            res.setEncoding('utf8');
            res.on('data', (chunk: string) => {
                buffer += chunk;
                const lines = buffer.split('\n');
                buffer = lines.pop() || '';

                let currentEvent = '';
                let currentDataLines: string[] = [];

                for (const line of lines) {
                    if (line.startsWith('event: ')) {
                        if (currentEvent && currentDataLines.length > 0) {
                            processEvent({ event: currentEvent, data: currentDataLines.join('\n') });
                        }
                        currentEvent = line.slice(7).trim();
                        currentDataLines = [];
                    } else if (line.startsWith('data: ')) {
                        currentDataLines.push(line.slice(6));
                    } else if (line.trim() === '' && currentEvent) {
                        if (currentDataLines.length > 0) {
                            processEvent({ event: currentEvent, data: currentDataLines.join('\n') });
                        }
                        currentEvent = '';
                        currentDataLines = [];
                    }
                }

                if (currentEvent && currentDataLines.length > 0) {
                    processEvent({ event: currentEvent, data: currentDataLines.join('\n') });
                    currentEvent = '';
                    currentDataLines = [];
                }
            });

            res.on('end', () => finish());
            res.on('error', (err) => finish(err));
        });

        req.on('error', (err) => finish(err));

        if (options.timeoutMs) {
            req.setTimeout(options.timeoutMs, () => {
                req.destroy(new Error('SSE request timeout'));
            });
        }

        if (options.body) {
            req.write(options.body);
        }
        req.end();

        function processEvent(evt: SseEvent) {
            result.raw_events.push(evt);
            if (options.onEvent) {
                options.onEvent(evt);
            }

            if (evt.event === 'DONE') {
                try {
                    const parsed = JSON.parse(evt.data);
                    if (parsed.conversation_id) {
                        result.conversation_id = parsed.conversation_id;
                    }
                } catch { /* ignore */ }
                finish();
                return;
            }

            if (evt.event === 'META') {
                try {
                    const parsed = JSON.parse(evt.data);
                    if (parsed.conversation) {
                        result.conversation = parsed.conversation;
                    }
                } catch { /* ignore */ }
                return;
            }

            if (evt.event === 'CREATE_STEP' || evt.event === 'SET_STATUS') {
                try {
                    const parsed = JSON.parse(evt.data);
                    if (evt.event === 'CREATE_STEP') {
                        result.steps.push({
                            id: parsed.id,
                            title: parsed.title,
                            status: 'CREATED',
                            tool_name: parsed.details?.tool_name
                        });
                    } else if (evt.event === 'SET_STATUS') {
                        const step = result.steps.find((s) => s.id === parsed.id);
                        if (step) {
                            step.status = parsed.status;
                            if (parsed.title) step.title = parsed.title;
                        }
                    }
                } catch { /* ignore */ }
                return;
            }

            if (evt.event === 'STEP_OUTPUT') {
                try {
                    const parsed = JSON.parse(evt.data);
                    const step = result.steps.find((s) => s.id === parsed.id);
                    if (step && step.tool_name === 'send_spl' && parsed.content) {
                        result.spl = extractSplFromMarkdown(parsed.content);
                    }
                } catch { /* ignore */ }
            }
        }
    });
}

function extractSplFromMarkdown(content: string): string {
    const match = content.match(/```spl\s*\n([\s\S]*?)\n```/);
    if (match) {
        return match[1].trim();
    }
    return content.replace(/```[\s\S]*?```/g, '').trim();
}
